src/Span.ts

```diff
@@ -0,0 +1,375 @@
+    if (carrier instanceof EventMessage || (('metadata' in carrier))) path = 'metadata.trace'
+    else if (('trace' in carrier)) path = 'trace'
+    else if (carrier instanceof EventTraceMetadata) return Promise.resolve(this.spanContext)
+    if (path) {
+      let pathArray: string[] = path.split('.')
+      for (let i = 0; i < pathArray.length - 1; i++) {
+        if (!result[pathArray[i]]) {
+          if (i <= pathArray.length) {
+            let o: any = {}
+            o[pathArray[i + 1]] = {}
+            result[pathArray[i]] = o
+          }
+        }
+        result = result[pathArray[i]]
+      }
+    }
+    result.trace = this.spanContext
+    return Promise.resolve(carrier)
+  }
+
+  /**
+   * Sets tags to the current span. If child span is created, the tags are passed on.
+   * @param tags key value pairs of tags. Tags can be changed on different child spans
+   */
+  setTags(tags: TraceTags): this {
+    let newContext: TypeSpanContext = new EventTraceMetadata(this.getContext())
+    for (let key in tags) {
+      newContext.tags![key] = tags[key]
+    }
+    this.spanContext = Object.freeze(new EventTraceMetadata(newContext))
+    return this
+  }
+
+  /**
+* Finishes the current span and its trace and sends the data to the tracing framework.
+* @param finishTimestamp optional parameter for the finish time. If omitted, current time is used.
+*/
+  async finish(message?: TypeOfMessage, finishTimestamp?: string | Date): Promise<this> {
+    if (this.spanContext.finishTimestamp) return Promise.reject(new Error('span already finished'))
+    let spanContext = this._finishSpan(finishTimestamp).getContext()
+    await this.trace(message, spanContext)
+    return Promise.resolve(this)
+  }
+
+  /**
+   * Finishes the trace by adding finish timestamp to the current span.
+   * @param finishTimestamp optional parameter for the finish time. If omitted, current time is used.
+   */
+  private _finishSpan(finishTimestamp?: string | Date): this {
+    let newContext: TypeSpanContext = <TypeSpanContext>Object.assign({}, this.spanContext)
+    if (finishTimestamp instanceof Date) {
+      newContext.finishTimestamp = finishTimestamp.toISOString() // ISO 8601
+    } else if (!finishTimestamp) {
+      newContext.finishTimestamp = (new Date()).toISOString() // ISO 8601
+    } else {
+      newContext.finishTimestamp = finishTimestamp
+    }
+    this.spanContext = Object.freeze(new EventTraceMetadata(newContext))
+    return this
+  }
+
+
+  /**
+   * Sends trace message to the tracing framework
+   * @param spanContext optional parameter. Can be used to trace previous span. If not set, the current span context is used.
+   * @param traceOptions options for status and event action. Default action is 'span' and status is success
+   */
+  private async trace(message?: TypeOfMessage, spanContext: TypeSpanContext = this.spanContext, action?: TraceEventAction, state?: EventStateMetadata): Promise<any> {
```

Comment:
Shouldn't you comment this method out since we are not using it at the moment?

Reply:
we are actually. the finish method is using it. I was actually planning to discuss with you should make it public since its okay and working

Reply:
That's cool. Let's leave it private for now.